aodndata/soop/soop_co2.py

```diff
@@ -0,0 +1,80 @@
+from aodncore.pipeline import HandlerBase, PipelineFile, PipelineFilePublishType, PipelineFileCheckType, FileType
+from aodncore.pipeline.exceptions import InvalidInputFileError
+import soop_co2_nrt_nc_generator
+from soop_co2_classifier import SoopCo2FileClassifier
+import re
+import os
+
+class SoopCo2Handler(HandlerBase):
+    """
+    Handler for data from IMOS SOOP CO2(DM, RT) and Future Reef Map Project.
+    It handles the following file types:
+         * ZIP;
+         * NetCDF;
+         * TXT;
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(SoopCo2Handler, self).__init__(*args, **kwargs)
+        self.allowed_extensions = ['.nc', '.txt', '.zip']
+
+    def preprocess(self):
+        """ Preprocessing for NRT and DM files
+           - NRT: generate a NetCDF files based on input text file.
+             Set the input file publish_type property to 'archive'
+           - DM file collection: update the check_type and publish_type properties for non-NetCDF files.
+             These files are not checked or harvested, but uploaded to S3
+
+        """
+        # Delayed mode file submitted as a zip archive
+        if self.file_extension == '.zip':
+            nc_file = self.file_collection.filter_by_attribute_id('file_type', FileType.NETCDF)
+            if len(nc_file) != 1:
+                raise InvalidInputFileError(
+                    "Expecting one netCDF file in ZIP archive '{zip}'".format(zip=os.path.basename(self.input_file))
+                )
+
+            # first process the NetCDF file to set the destination path for the file collection
+            nc = nc_file[0]
+            set_nc_dest_path = SoopCo2FileClassifier.dest_path(nc)
+            nc.dest_path = os.path.join(set_nc_dest_path, nc.name)
+            nc.publish_type = PipelineFilePublishType.HARVEST_UPLOAD
+            nc.check_type = PipelineFileCheckType.NC_COMPLIANCE_CHECK
+
+            # SOOP-CO2 DM and FRMAP .txt,.pdf or/and .xml files. Set check type to NONEMPTY and publish type to UPLOAD_ONLY
+            non_nc_files = self.file_collection.filter_by_attribute_id('check_type', PipelineFileCheckType.NO_ACTION)
+            for non_nc in non_nc_files:
+                non_nc.check_type = PipelineFileCheckType.FORMAT_CHECK
+                non_nc.publish_type = PipelineFilePublishType.UPLOAD_ONLY
+                non_nc.dest_path = os.path.join(set_nc_dest_path, non_nc.name)
+
+        elif self.input_file.endswith('dat.txt'):
```

Comment:
```You don't need to assign the dest_path attribute to the file or call the dest_path function yourself. This should be calculated and assigned automatically later on during the publish step, based on the fact that you added the following below```
I call the dest_path function in case when I need to work out the destination path for ancillary files that don't carry any/enough information for the dest_path function to calculate it directly.

Reply:
and Thanks for your input @lwgordonimos  👍

Reply:
'''The dest_path function/method should return the entire path, including the filename.'''
Except that it's not convenient to re-use the dest path if it includes the (netcdf) filename.

Reply:
Ah yes, that rings a bell. Makes sense. I have been thinking a bit about the dest_path stuff and think this sort of thing could be made cleaner by:

1. being able to add some custom metadata to a PipelineFile
1. making the dest_path functions take a PipelineFile object as a parameter (instead of just a path)

This would mean the dest_path functions would potentially have a lot more information to work with. But definitely a "future enhancement".

Reply:
Are you saying I should leave the bits where I process ancillary files off the pipeline until the enhancement is implemented?

Reply:
Sorry, wasn't clear. No, I'm saying to do exactly what you're doing here now, because it's the only (and currently best) way, but that future enhancement idea could make this simpler.

This way is perfectly valid, and is in fact why I added this check https://github.com/aodn/python-aodncore/blob/master/aodncore/pipeline/files.py#L693, so that if a handler author wanted to set it manually, the publish step would *not* try to overwrite it, because it assumes it was set earlier  for a reason :)

Reply:
'''PipelineFilePublishType.HARVEST_UPLOAD and PipelineFileCheckType.NC_COMPLIANCE_CHECK are the defaults for NetCDF files, so you don't need to explicitly assign them here'''
Is this suppose to apply also to product? As soon as you create a netcdf file it these attribute are assigned by default? Cause this is not the behaviour i see when running the unittest

Reply:
No you're right, not for products. By default they are NO_ACTION, so you can set them to whatever.

I should have put the comment against the actual line:
https://github.com/aodn/python-aodndata/pull/10/files/5d95d90e833eaccc919f6cbf3ce1086fa1966304#diff-c9a870baa45ba537fc28ab45791fdc2bR69